server/app/services/program/predicate/PredicateGenerator.java

```diff
@@ -0,0 +1,290 @@
+              "Missing or unknown predicateAction: %s", predicateForm.get("predicateAction")));
+    }
+
+    for (String key : predicateForm.rawData().keySet()) {
+      Matcher singleValueMatcher = SINGLE_PREDICATE_VALUE_FORM_KEY_PATTERN.matcher(key);
+      Matcher multiValueMatcher = MULTI_PREDICATE_VALUE_FORM_KEY_PATTERN.matcher(key);
+
+      int groupId;
+      long questionId;
+      Scalar scalar;
+      Operator operator;
+      PredicateValue predicateValue;
+
+      if (singleValueMatcher.find()) {
+        Pair<Integer, Long> groupIdQuestionIdPair =
+            getGroupIdAndQuestionId(roQuestionService, singleValueMatcher);
+        groupId = groupIdQuestionIdPair.getLeft();
+        questionId = groupIdQuestionIdPair.getRight();
+        Pair<Scalar, Operator> scalarOperatorPair = getScalarAndOperator(predicateForm, questionId);
+        scalar = scalarOperatorPair.getLeft();
+        operator = scalarOperatorPair.getRight();
+
+        predicateValue =
+            parsePredicateValue(scalar, operator, predicateForm.get(key), ImmutableList.of());
+      } else if (multiValueMatcher.find() && !consumedMultiValueKeys.contains(key)) {
+        Pair<Integer, Long> groupIdQuestionIdPair =
+            getGroupIdAndQuestionId(roQuestionService, multiValueMatcher);
+        groupId = groupIdQuestionIdPair.getLeft();
+        questionId = groupIdQuestionIdPair.getRight();
+        Pair<Scalar, Operator> scalarOperatorPair = getScalarAndOperator(predicateForm, questionId);
+        scalar = scalarOperatorPair.getLeft();
+        operator = scalarOperatorPair.getRight();
+
+        ImmutableList<String> multiSelectKeys =
+            predicateForm.rawData().keySet().stream()
+                .filter(
+                    filteredKey ->
+                        filteredKey.startsWith(
+                            String.format(
+                                "group-%d-question-%d-predicateValues", groupId, questionId)))
+                .collect(ImmutableList.toImmutableList());
+
+        consumedMultiValueKeys.addAll(multiSelectKeys);
+
+        ImmutableList<String> rawPredicateValues =
+            multiSelectKeys.stream()
+                .map(predicateForm.rawData()::get)
+                .collect(ImmutableList.toImmutableList());
+
+        predicateValue = parsePredicateValue(scalar, operator, "", rawPredicateValues);
+      } else {
+        continue;
+      }
+
+      leafNodes.put(
+          groupId,
+          LeafOperationExpressionNode.builder()
+              .setQuestionId(questionId)
+              .setScalar(scalar)
+              .setOperator(operator)
+              .setComparedValue(predicateValue)
+              .build());
+    }
+
+    switch (detectFormat(leafNodes)) {
+      case OR_OF_SINGLE_LAYER_ANDS:
+        {
+          return PredicateDefinition.create(
+              PredicateExpressionNode.create(
+                  OrNode.create(
+                      leafNodes.keySet().stream()
```

Comment:
I think you need to order the hashmap's keyset to maintain the creation order in the UI.  I don't see how that's happening otherwise.

Reply:
leafNodes is a [LinkedHashMultimap](https://guava.dev/releases/19.0/api/docs/com/google/common/collect/LinkedHashMultimap.html) which has stable order:

> The collections returned by keySet, keys, and asMap iterate through the keys in the order they were first added to the multimap. 

To my dismay, the underlying Map in play's Form/DynamicForm classes is a simple HashMap with no guaranteeing of ordering. I'll add ordering to the top-level loop.

Reply:
I take it back, since the leaf nodes are keyed in the multimap by group ID we can sort on that ID to ensure the ordering is the same as presented in the UI

Reply:
Yeah, was suggesting ordering by the group key.  I don't think we'd want to rely on the insertion order as that implies we process the form keys in a deterministic and "correct for our usage" order which I don't imagine it'd provide us.